pkg/spdk/server.go

```diff
@@ -71,59 +71,83 @@ func (s *Server) monitoringReplicas() {
 			logrus.Info("SPDK Server: stopped monitoring replicas due to the context done")
 			done = true
 		case <-ticker.C:
-			bdevLvolList, err := s.spdkClient.BdevLvolGet("", 0)
-			if err != nil {
-				logrus.Errorf("SPDK Server: failed to get lvol bdevs for replica cache update, will retry later: %v", err)
-				continue
+			if err := s.verifyReplicas; err != nil {
+				logrus.Errorf("SPDK Server: failed to verify and update replica cache, will retry later: %v", err)
 			}
-			bdevLvolMap := map[string]*spdktypes.BdevInfo{}
-			for idx := range bdevLvolList {
-				bdevLvol := &bdevLvolList[idx]
-				if len(bdevLvol.Aliases) == 0 {
-					continue
-				}
-				bdevLvolMap[spdktypes.GetLvolNameFromAlias(bdevLvol.Aliases[0])] = bdevLvol
-			}
-
-			s.Lock()
-			for _, r := range s.replicaMap {
-				r.ValidateAndUpdate(bdevLvolMap)
-			}
-			s.Unlock()
 		}
 		if done {
 			break
 		}
 	}
 }
 
+func (s *Server) verifyReplicas() error {
+	s.Lock()
+	defer s.Unlock()
+
+	bdevLvolList, err := s.spdkClient.BdevLvolGet("", 0)
+	if err != nil {
+		return err
+	}
+	subsystemList, err := s.spdkClient.NvmfGetSubsystems("", "")
+	if err != nil {
+		return err
+	}
+	bdevLvolMap := map[string]*spdktypes.BdevInfo{}
+	for idx := range bdevLvolList {
+		bdevLvol := &bdevLvolList[idx]
+		if len(bdevLvol.Aliases) == 0 {
+			continue
+		}
+		bdevLvolMap[spdktypes.GetLvolNameFromAlias(bdevLvol.Aliases[0])] = bdevLvol
+	}
+	subsystemMap := map[string]*spdktypes.NvmfSubsystem{}
+	for idx := range subsystemList {
+		subsystem := &subsystemList[idx]
+		subsystemMap[subsystem.Nqn] = subsystem
+	}
+
+	for _, r := range s.replicaMap {
+		r.ValidateAndUpdate(s.spdkClient, bdevLvolMap, subsystemMap)
+	}
+
+	return nil
+}
+
 func (s *Server) ReplicaCreate(ctx context.Context, req *spdkrpc.ReplicaCreateRequest) (ret *spdkrpc.Replica, err error) {
+	if req.Name == "" {
+		return nil, grpcstatus.Error(grpccodes.InvalidArgument, "replica name is required")
+	}
+	if req.LvsName == "" || req.LvsUuid == "" {
+		return nil, grpcstatus.Error(grpccodes.InvalidArgument, "lvs name and lvs UUID are required")
+	}
+
 	s.Lock()
+	defer s.Unlock()
+
 	if _, ok := s.replicaMap[req.Name]; ok {
-		s.Unlock()
 		return nil, grpcstatus.Errorf(grpccodes.AlreadyExists, "replica %v already exists", req.Name)
 	}
 
 	s.replicaMap[req.Name] = NewReplica(req.Name, req.LvsName, req.LvsUuid, req.SpecSize)
 	r := s.replicaMap[req.Name]
-	s.Unlock()
 
-	portStart, _, err := s.portAllocator.AllocateRange(1)
-	if err != nil {
-		return nil, err
-	}
-
-	return r.Create(s.spdkClient, portStart, req.ExposeRequired)
+	return r.Create(s.spdkClient, req.ExposeRequired, s.portAllocator)
```

Comment:
Remove `r` from `replicaMap` if it is unable to be created.

Reply:
The current strategy is that retaining invalid/error replicas then the caller is responsible for cleaning them up

Reply:
Got it. If the creation fails, I prefer to clean it up before exiting the function, but I'm fine with the implementation.

Reply:
Another question, does legacy volume replica also adopt the same strategy? If no, will we need a different error handling logic in longhorn-manager?